### PR TITLE
new Orch API test: PublishedWorkspaceSpec

### PR DESCRIPTION
**Pull request by aweng98**
[QA-477](https://broadworkbench.atlassian.net/browse/QA-477), [QA-478](https://broadworkbench.atlassian.net/browse/QA-478), [QA-479](https://broadworkbench.atlassian.net/browse/QA-479)
depend on workbench-lib PR [223](https://github.com/broadinstitute/workbench-libs/pull/223)

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/PublishedWorkspaceSpec.scala [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  /**
+    *
+    * @return True: workspace is visible in library table
+    */
+  private def isVisibleInLibrary(namespace: String, workspaceName: String)(implicit token: AuthToken): Boolean = {
+    val response = Orchestration.library.searchPublishedLibraryDataset(namespace, workspaceName)
+
+    import DefaultJsonProtocol._
+    val total = JsonParser(response).asJsObject.fields("total").convertTo[Int]
+    total == 1
```
Comment (marctalbott) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/PublishedWorkspaceSpec.scala:172]: Could Orch ever return more than one result for this search and should the test fail if it does?

**Comment by aweng98** [comment]
@marctalbott @ansingh7115 I can add "Search library by research purpose" tests [QA-475](https://broadworkbench.atlassian.net/browse/QA-475) in this `spec` easily. What do you think? If not to add here, this PR is ready for review.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/PublishedWorkspaceSpec.scala [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        "should default to visible to 'all users' in Library Attributes" in {
+          val curatorUser = UserPool.chooseCurator
+          implicit val curatorAuthToken: AuthToken = curatorUser.makeAuthToken()
+
+          withCleanBillingProject(curatorUser) { billingProject =>
+            withWorkspace(billingProject, "PublishedWorkspaceSpec_workspace") { workspaceName =>
+
+              val data = LibraryData.metadataBasic + ("library:datasetName" -> workspaceName)
+              Orchestration.library.setLibraryAttributes(billingProject, workspaceName, data)
+              Orchestration.library.setDiscoverableGroups(billingProject, workspaceName, List("all_broad_users"))
```
Comment (marctalbott) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/PublishedWorkspaceSpec.scala:95]: This needs to be set? By setting this doesn't it change the test to not be checking the "default" behavior?

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/PublishedWorkspaceSpec.scala (hunk 1 above)
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/PublishedWorkspaceSpec.scala:172]: no. bc param search string is the `workspaceName`.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/PublishedWorkspaceSpec.scala (hunk 2 above)
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/PublishedWorkspaceSpec.scala:95]: yes, it is needed. the default, `discovereableGroup` is not set or has any value. cloned workspace cannot inherit property `getDiscoverableGroup` from the workspace used in clone. that's verified at line 102.

**Comment by aweng98** [comment]
jenkins retest

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/PublishedWorkspaceSpec.scala [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+          withWorkspace(billingProject, "PublishedWorkspaceSpec_workspace") { workspaceName =>
+
+            val data = LibraryData.metadataBasic + ("library:datasetName" -> workspaceName)
+            Orchestration.library.setLibraryAttributes(billingProject, workspaceName, data)
+            Orchestration.library.publishWorkspace(billingProject, workspaceName)
+
+            withClue("a published workspace should be visible in library.") {
+              eventually {
+                isVisibleInLibrary(workspaceName) shouldBe true
+              }
```
Comment (ansingh7115) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/PublishedWorkspaceSpec.scala:37]: Correct me if I'm but wrong, but I don't believe this requires an `eventually` - publishWorkspace returns after the workspace has been published, which means it should immediately be available to search - same with other places `isVisibleInLibrary` is used.

Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/PublishedWorkspaceSpec.scala:37]: I saw few times where it took 2 calls to search library to get expected result.  `eventually` will help reduce test flakiness.

File: automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/PublishedWorkspaceSpec.scala [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+                }
+              }
+            }
+          }
+        }
+
+      }
+    }
+
+    "For a dataset with consent codes" in {
```
Comment (marctalbott) [inline automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/PublishedWorkspaceSpec.scala]: This test name and the name of the next test should be more descriptive of what the tests are doing. Currently, the names of the tests are `For a user with publish permissions For a dataset with consent codes` and `For a user with publish permissions For a dataset with tags` which don't tell me much about the test. I know these are the names of the tests already, but it'd be great if we could rename them to `A user with publish permissions can publish a dataset with consent codes` or something along those lines. I may also be misunderstanding the test so I'm not sure if my suggested name change is a good one

Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/test/api/orch/PublishedWorkspaceSpec.scala]: 👍 updated names. let me know if new names still confusing.

**Review by marctalbott — APPROVED** [review]

**Comment by aweng98** [comment]
jenkins retest

**Review by ansingh7115 — APPROVED** [review]

**Comment by aweng98** [comment]
fiab start failed 😔
jenkins retest

**Pull request merged**